gtk4/src/subclass/widget.rs

```diff
@@ -983,6 +983,75 @@ pub unsafe trait WidgetClassSubclassExt: ClassStruct {
         }
     }
 
+    #[doc(alias = "gtk_widget_class_add_binding")]
+    fn add_binding<
+        F: Fn(&<<Self as ClassStruct>::Type as ObjectSubclass>::Type, Option<&Variant>) -> bool
+            + 'static,
+    >(
+        &mut self,
+        keyval: gdk::keys::Key,
+        mods: gdk::ModifierType,
+        callback: F,
+        arguments: Option<&glib::Variant>,
+    ) {
+        let shortcut = crate::Shortcut::new(
+            Some(&crate::KeyvalTrigger::new(keyval, mods)),
+            Some(&crate::CallbackAction::new(move |widget, args| -> bool {
+                unsafe { callback(widget.unsafe_cast_ref(), args) }
+            })),
+        );
+        shortcut.set_arguments(arguments);
+        self.add_shortcut(&shortcut);
+    }
+
+    #[doc(alias = "gtk_widget_class_add_binding_action")]
+    fn add_binding_action(
+        &mut self,
+        keyval: gdk::keys::Key,
+        mods: gdk::ModifierType,
+        action_name: &str,
+        arguments: Option<&glib::Variant>,
+    ) {
+        let shortcut = crate::Shortcut::new(
+            Some(&crate::KeyvalTrigger::new(keyval, mods)),
+            Some(&crate::NamedAction::new(action_name)),
+        );
+        shortcut.set_arguments(arguments);
+        self.add_shortcut(&shortcut);
+    }
+
+    #[doc(alias = "gtk_widget_class_add_binding_signal")]
+    fn add_binding_signal(
+        &mut self,
+        keyval: gdk::keys::Key,
+        mods: gdk::ModifierType,
+        signal_name: &str,
+        arguments: Option<&glib::Variant>,
```

Comment:
Shouldn't this get some checks to see whether the signal exists and has the correct parameter type(s)?

Reply:
Ideally yes, but upstream doesn't neither and just has `/* XXX: validate variant format for signal */` same for the actions. I will see how doable that's

Reply:
The problem I see here is that this can easily cause memory corruption if it's not checked on the C side either.

Reply:
What's a bit difficult here is how to compare the types that you can get from a `SignalQuery::param_types` (returns a `Vec<SignalType>`) and `Variant::type_` (returns a `VariantTy`). Any ideas how to do that?

Reply:
How is the C code converting between them?